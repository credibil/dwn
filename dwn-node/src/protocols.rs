//! # Protocols

mod configure;
pub(crate) mod integrity;
mod query;

use std::collections::BTreeMap;
use std::sync::LazyLock;

use serde::{Deserialize, Serialize};

pub use self::configure::{
    Action, ActionRule, Actor, Configure, ConfigureDescriptor, ConfigureReply, Definition,
    ProtocolType, RuleSet, validate_structure,
};
pub use self::query::{Query, QueryBuilder, QueryReply};
use crate::Range;

/// Default protocol for managing web node permission grants.
pub const PROTOCOL_URI: &str = "https://vercre.website/dwn/permissions";

/// The protocol path of the `request` record.
pub const REQUEST_PATH: &str = "request";

/// The protocol path of the `grant` record.
pub const GRANT_PATH: &str = "grant";

///The protocol path of the `revocation` record.
pub const REVOCATION_PATH: &str = "grant/revocation";

/// Protocol filter.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProtocolsFilter {
    /// Protocol matching the specified protocol.
    pub protocol: String,
}

/// Default protocol definition.
pub static DEFINITION: LazyLock<Definition> = LazyLock::new(|| {
    // default types
    let mut types = BTreeMap::new();
    let default_type = ProtocolType {
        data_formats: Some(vec!["application/json".to_string()]),
        ..ProtocolType::default()
    };
    types.insert("request".to_string(), default_type.clone());
    types.insert("grant".to_string(), default_type.clone());
    types.insert("revocation".to_string(), default_type);

    // default structure (aka rules)
    let default_size = Range {
        min: None,
        max: Some(10000),
    };

    let mut structure = BTreeMap::new();
    structure.insert("request".to_string(), RuleSet {
        size: Some(default_size.clone()),
        actions: Some(vec![ActionRule {
            who: Some(Actor::Anyone),
            can: vec![Action::Create],
            ..ActionRule::default()
        }]),
        ..RuleSet::default()
    });
    structure.insert("grant".to_string(), RuleSet {
        size: Some(default_size.clone()),
        actions: Some(vec![ActionRule {
            who: Some(Actor::Recipient),
            of: Some("grant".to_string()),
            can: vec![Action::Read, Action::Query],
            ..ActionRule::default()
        }]),
        // revocation is nested under grant
        structure: BTreeMap::from([("revocation".to_string(), RuleSet {
            size: Some(default_size),
            actions: Some(vec![ActionRule {
                who: Some(Actor::Anyone),
                can: vec![Action::Read],
                ..ActionRule::default()
            }]),
            ..RuleSet::default()
        })]),
        ..RuleSet::default()
    });

    Definition {
        protocol: PROTOCOL_URI.to_string(),
        published: true,
        types,
        structure,
    }
});
