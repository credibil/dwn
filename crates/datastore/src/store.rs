//! # Store
//!
//! The `store` module provides utilities for storing and retrieving messages
//! and associated data.
//!
//! The two primary types exposed by this module are [`Storable`] and [`Query`].
//!
//! [`Storable`] wraps each message with a unifying type used to simplify storage
//! and retrieval as well as providing a vehicle for attaching addtional data
//! alongside the message (i.e. indexes).

use anyhow::{Result, anyhow};
use serde::Serialize;
use serde::de::DeserializeOwned;

use crate::query::{Cursor, Query};
use crate::{BlockStore, Document, Storable, index, ipfs};

/// Store a message in the underlying store.
pub async fn put(
    owner: &str, partition: &str, entry: &impl Storable, store: &impl BlockStore,
) -> Result<()> {
    let document = entry.document();

    // store entry block
    let message_cid = document.cid()?;
    store.delete(owner, partition, &message_cid).await?;
    store.put(owner, partition, &message_cid, &ipfs::encode_block(&document)?).await?;

    // index entry
    index::insert(owner, partition, entry, store).await
}

/// Queries the underlying store for matches to the provided query.
pub async fn query<T>(
    owner: &str, partition: &str, query: &Query, store: &impl BlockStore,
) -> Result<(Vec<T>, Option<Cursor>)>
where
    T: Serialize + DeserializeOwned,
{
    let mut results = index::query(owner, partition, query, store).await?;

    // return cursor when paging is used
    let limit = query.pagination.as_ref().map_or(0, |p| p.limit.unwrap_or(0));
    let cursor = if limit > 0 && limit < results.len() {
        let sort_field = query.sort.to_string();

        // set cursor to the last item remaining after the spliced result.
        results.pop().map(|item| Cursor {
            message_cid: item.message_cid.clone(),
            value: item.fields[&sort_field].clone(),
        })
    } else {
        None
    };

    let mut entries = Vec::new();
    for item in results {
        let Some(bytes) = store.get(owner, partition, &item.message_cid).await? else {
            return Err(anyhow!("missing block for message cid"));
        };
        entries.push(ipfs::decode_block(&bytes)?);
    }

    Ok((entries, cursor))
}

/// Fetch a single message by CID from the underlying store, returning
/// `None` if no message was found.
pub async fn get<T>(
    owner: &str, partition: &str, message_cid: &str, store: &impl BlockStore,
) -> Result<Option<T>>
where
    T: Serialize + DeserializeOwned,
{
    let Some(bytes) = store.get(owner, partition, message_cid).await? else {
        return Ok(None);
    };
    Ok(Some(ipfs::decode_block(&bytes)?))
}

/// Delete message associated with the specified id.
pub async fn delete(
    owner: &str, partition: &str, message_cid: &str, store: &impl BlockStore,
) -> Result<()> {
    index::delete(owner, partition, message_cid, store).await?;
    store.delete(owner, partition, message_cid).await
}
