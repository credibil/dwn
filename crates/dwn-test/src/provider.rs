#![allow(missing_docs)]
#![allow(unused_variables)]

//! # Provider
//!
//! Implementation of the `Provider` trait for testing and examples.

pub mod block_store;
pub mod event_log;
mod event_stream;
pub mod key_store;
pub mod message_store;
pub mod task_store;

use anyhow::{Result, anyhow};
use blockstore::InMemoryBlockstore;
use surrealdb::Surreal;
use surrealdb::engine::local::{Db, Mem};
use vercre_dwn_server::provider::{DidResolver, Document, Provider};

use self::key_store::{ALICE_DID, KeyStoreImpl};

const NAMESPACE: &str = "integration-test";

#[derive(Clone)]
pub struct ProviderImpl {
    db: Surreal<Db>,
    blockstore: InMemoryBlockstore<64>,
    pub nats_client: async_nats::Client,
    pub keystore: KeyStoreImpl,
}

impl Provider for ProviderImpl {}

impl ProviderImpl {
    pub async fn new() -> Result<Self> {
        let db = Surreal::new::<Mem>(()).await?;
        db.use_ns(NAMESPACE).use_db(ALICE_DID).await?;

        Ok(Self {
            db,
            blockstore: InMemoryBlockstore::<64>::new(),
            nats_client: async_nats::connect("demo.nats.io").await?,
            keystore: KeyStoreImpl::new(),
        })
    }
}

impl DidResolver for ProviderImpl {
    async fn resolve(&self, url: &str) -> Result<Document> {
        if url == ALICE_DID {
            return serde_json::from_slice(include_bytes!("./provider/data/alice_did.json"))
                .map_err(|e| anyhow!(format!("issue deserializing document: {e}")));
        } else {
            return serde_json::from_slice(include_bytes!("./provider/data/bob_did.json"))
                .map_err(|e| anyhow!(format!("issue deserializing document: {e}")));
        }
    }
}
