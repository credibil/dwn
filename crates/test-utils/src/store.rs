//! # In-Memory Datastore

use std::sync::LazyLock;

use anyhow::Result;
// use credibil_core::datastore::Datastore;
use credibil_ecc::Vault;
use dashmap::DashMap;

static STORE: LazyLock<DashMap<String, Vec<u8>>> = LazyLock::new(DashMap::new);

#[derive(Clone, Debug)]
pub struct Datastore;

impl Datastore {
    pub async fn put(owner: &str, partition: &str, key: &str, data: &[u8]) -> Result<()> {
        let key = format!("{owner}-{partition}-{key}");
        STORE.insert(key, data.to_vec());
        Ok(())
    }

    pub async fn get(owner: &str, partition: &str, key: &str) -> Result<Option<Vec<u8>>> {
        let key = format!("{owner}-{partition}-{key}");
        let Some(bytes) = STORE.get(&key) else {
            return Ok(None);
        };
        Ok(Some(bytes.to_vec()))
    }

    pub async fn delete(owner: &str, partition: &str, key: &str) -> Result<()> {
        let key = format!("{owner}-{partition}-{key}");
        STORE.remove(&key);
        Ok(())
    }

    pub async fn get_all(owner: &str, partition: &str) -> Result<Vec<(String, Vec<u8>)>> {
        let all = STORE
            .iter()
            .filter(move |r| r.key().starts_with(&format!("{owner}-{partition}-")))
            .map(|r| (r.key().to_string(), r.value().clone()))
            .collect::<Vec<_>>();
        Ok(all)
    }
}

#[derive(Clone, Debug)]
pub struct KeyVault;

impl Vault for KeyVault {
    async fn put(&self, owner: &str, partition: &str, key: &str, data: &[u8]) -> Result<()> {
        Datastore::put(owner, partition, key, data).await
    }

    async fn get(&self, owner: &str, partition: &str, key: &str) -> Result<Option<Vec<u8>>> {
        Datastore::get(owner, partition, key).await
    }

    async fn delete(&self, owner: &str, partition: &str, key: &str) -> Result<()> {
        Datastore::delete(owner, partition, key).await
    }

    async fn get_all(&self, owner: &str, partition: &str) -> Result<Vec<(String, Vec<u8>)>> {
        Datastore::get_all(owner, partition).await
    }
}
