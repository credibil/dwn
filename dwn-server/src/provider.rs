//! # Provider

use anyhow::{Result, anyhow};
pub use vercre_did::{DidResolver, Document};
pub use vercre_infosec::{KeyOps, Receiver, Signer};

use crate::event::{Event, Subscriber};
use crate::store::Cursor;
pub use crate::store::{Entry, Query};
pub use crate::tasks::ResumableTask;

/// Issuer Provider trait.
pub trait Provider:
    MessageStore + BlockStore + TaskStore + EventLog + EventStream + KeyStore + DidResolver
{
}

/// The `KeyStore` trait is used to provide methods needed for signing,
/// encrypting, verifying, and decrypting data.
///
/// Implementers of this trait are expected to provide the necessary
/// cryptographic functionality to support Verifiable Credential issuance and
/// Verifiable Presentation submissions.
pub trait KeyStore: Send + Sync {
    /// Signer provides digital signing function.
    ///
    /// The `controller` parameter uniquely identifies the controller of the
    /// private key used in the signing operation.
    ///
    /// # Errors
    ///
    /// Returns an error if the signer cannot be created.
    fn keyring(&self, controller: &str) -> Result<impl Keyring>;

    // /// Signer provides digital signing function.
    // ///
    // /// The `controller` parameter uniquely identifies the controller of the
    // /// private key used in the signing operation.
    // ///
    // /// # Errors
    // ///
    // /// Returns an error if the signer cannot be created.
    // fn signer(&self, controller: &str) -> impl Future<Output = Result<impl Signer>>+Send;

    // /// Cipher provides data encryption/decryption functionality.
    // ///
    // /// The `controller` parameter uniquely identifies the controller of the
    // /// private key used in the signing operation.
    // ///
    // /// # Errors
    // ///
    // /// Returns an error if the encryptor cannot be created.
    // fn cipher(&self, controller: &str) -> impl Future<Output = Result<impl Cipher>>+Send;
}

/// The `Keyring` trait provides the methods needed for signing, encrypting,
/// verifying, and decrypting data.
///
/// Implementers of this trait are expected to provide the necessary
/// cryptographic functionality to support Verifiable Credential issuance and
/// Verifiable Presentation submissions.
pub trait Keyring: Signer + Receiver {}

/// The `MessageStore` trait is used by implementers to provide message
/// storage capability.
pub trait MessageStore: Send + Sync {
    /// Store a message in the underlying store.
    fn put(&self, owner: &str, record: &Entry) -> impl Future<Output = Result<()>> + Send;

    /// Queries the underlying store for matches to the provided query.
    fn query(&self, owner: &str, query: &Query) -> impl Future<Output = Result<Vec<Entry>>> + Send;

    /// Queries the underlying store returning the current page of results.
    fn paginated_query(
        &self, owner: &str, query: &Query,
    ) -> impl Future<Output = Result<(Vec<Entry>, Option<Cursor>)>> + Send {
        async move {
            let entries = self.query(owner, query).await?;

            // no pagination
            let Query::Records(query) = query else {
                return Ok((entries, None));
            };
            // no pagination
            let Some(pagination) = &query.pagination else {
                return Ok((entries, None));
            };
            // sorted on?
            let Some(sort_field) = &query.sort else {
                return Ok((entries, None));
            };
            // additional results?
            let mut limit = pagination.limit.unwrap_or(entries.len());
            if entries.len() <= limit {
                return Ok((entries, None));
            }

            // page offset
            let offset = if let Some(cursor) = &pagination.cursor {
                // find starting point
                let Some(index) =
                    entries.iter().position(|e| e.cid().ok().as_ref() == Some(&cursor.message_cid))
                else {
                    return Err(anyhow!("cursor `message_cid` is invalid"));
                };
                index
            } else {
                0
            };

            // don't blow upper bound
            if offset + limit > entries.len() {
                limit = entries.len() - offset;
            }

            // capture page
            let curr_page = entries.as_slice()[offset..offset + limit].to_vec();

            // starting point for the next page (using `message_cid`)
            let Some(next_entry) = entries.get(offset + limit) else {
                return Ok((curr_page, None));
            };

            Ok((
                curr_page,
                Some(Cursor {
                    message_cid: next_entry.cid()?,
                    sort_value: sort_field.to_string(),
                }),
            ))
        }
    }

    /// Fetches a single message by CID from the underlying store, returning
    /// `None` if no message was found.
    fn get(
        &self, owner: &str, message_cid: &str,
    ) -> impl Future<Output = Result<Option<Entry>>> + Send;

    /// Delete message associated with the specified id.
    fn delete(&self, owner: &str, message_cid: &str) -> impl Future<Output = Result<()>> + Send;

    /// Purge all records from the store.
    fn purge(&self) -> impl Future<Output = Result<()>> + Send;
}

/// The `BlockStore` trait is used by implementers to provide data storage
/// capability.
pub trait BlockStore: Send + Sync {
    /// Store a data block in the underlying block store.
    fn put(&self, owner: &str, cid: &str, block: &[u8]) -> impl Future<Output = Result<()>> + Send;

    /// Fetches a single block by CID from the underlying store, returning
    /// `None` if no match was found.
    fn get(&self, owner: &str, cid: &str) -> impl Future<Output = Result<Option<Vec<u8>>>> + Send;

    /// Delete the data block associated with the specified CID.
    fn delete(&self, owner: &str, cid: &str) -> impl Future<Output = Result<()>> + Send;

    /// Purge all blocks from the store.
    fn purge(&self) -> impl Future<Output = Result<()>> + Send;
}

/// The `TaskStore` trait is used by implementers to provide data storage
/// capability.
pub trait TaskStore: Send + Sync {
    /// Registers a new resumable task that is currently in-flight/under
    /// processing to the store.
    ///
    /// If the task has timed out, a client will be able to grab it through the
    /// `grab()` method and resume the task.
    fn register(
        &self, owner: &str, task: &ResumableTask, timeout_secs: u64,
    ) -> impl Future<Output = Result<()>> + Send;

    /// Grabs `count` unhandled tasks from the store.
    ///
    /// Unhandled tasks are tasks that are not currently in-flight/under processing
    /// (ie. tasks that have timed-out).
    ///
    /// N.B.: The implementation must make sure that once a task is grabbed by a client,
    /// tis timeout must be updated so that it is considered in-flight/under processing
    /// and cannot be grabbed by another client until it is timed-out.
    fn grab(
        &self, owner: &str, count: u64,
    ) -> impl Future<Output = Result<Vec<ResumableTask>>> + Send;

    /// Reads the task associated with the task ID provided regardless of whether
    /// it is in-flight/under processing or not.
    ///
    /// This is mainly introduced for testing purposes: ie. to check the status of
    /// a task for easy test verification.
    fn read(
        &self, owner: &str, task_id: &str,
    ) -> impl Future<Output = Result<Option<ResumableTask>>> + Send;

    /// Extends the timeout of the task associated with the task ID provided.
    ///
    /// No-op if the task is not found, as this implies that the task has already
    /// been completed. This allows the client that is executing the task to
    /// continue working on it before the task is considered timed out.
    fn extend(
        &self, owner: &str, task_id: &str, timeout_secs: u64,
    ) -> impl Future<Output = Result<()>> + Send;

    /// Delete data associated with the specified id.
    fn delete(&self, owner: &str, task_id: &str) -> impl Future<Output = Result<()>> + Send;

    /// Purge all data from the store.
    fn purge(&self, owner: &str) -> impl Future<Output = Result<()>> + Send;
}

/// The `Metadata` trait is used by implementers to provide `Client`, `Issuer`,
/// and `Server` metadata to the library.
pub trait EventLog: Send + Sync {
    /// Adds a message event to a owner's event log.
    fn append(&self, owner: &str, event: &Event) -> impl Future<Output = Result<()>> + Send;

    /// Retrieves all of a owner's events that occurred after the cursor provided.
    /// If no cursor is provided, all events for a given owner will be returned.
    ///
    /// The cursor is a `message_cid`.
    fn events(
        &self, owner: &str, cursor: Option<Cursor>,
    ) -> impl Future<Output = Result<(Vec<Event>, Cursor)>> + Send;

    /// Retrieves a filtered set of events that occurred after a the cursor
    /// provided, accepts multiple filters. If no cursor is provided, all
    /// events for a given owner and filter combo will be returned. The cursor
    /// is a `message_cid`.
    ///
    /// Returns an array of `message_cid`s that represent the events.
    fn query(
        &self, owner: &str, query: &Query,
    ) -> impl Future<Output = Result<(Vec<Event>, Cursor)>> + Send;

    /// Deletes event for the specified `message_cid`.
    fn delete(&self, owner: &str, message_cid: &str) -> impl Future<Output = Result<()>> + Send;

    /// Purge all data from the store.
    fn purge(&self) -> impl Future<Output = Result<()>> + Send;
}

/// The `EventStream` trait is used by implementers to provide `Client`, `Issuer`,
/// and `Server` metadata to the library.
pub trait EventStream: Send + Sync {
    /// Subscribes to an owner's event stream.
    fn subscribe(&self, owner: &str) -> impl Future<Output = Result<Subscriber>> + Send;

    /// Emits an event to a owner's event stream.
    fn emit(&self, owner: &str, event: &Event) -> impl Future<Output = Result<()>> + Send;
}
