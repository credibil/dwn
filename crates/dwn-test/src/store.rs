#![allow(missing_docs)]
#![allow(unused_variables)]

//! # Provider
//!
//! Implementation of the `Provider` trait for testing and examples.

pub mod block;
pub mod event;
pub mod message;
pub mod task;

use anyhow::{anyhow, Result};
use blockstore::InMemoryBlockstore;
use serde::Deserialize;
use surrealdb::engine::local::{Db, Mem};
use surrealdb::opt::RecordId;
use surrealdb::Surreal;
use vercre_dwn::protocols::Configure;
use vercre_dwn::provider::{DidResolver, Document, MessageStore, Provider};

use crate::keystore::{KeystoreImpl, ALICE_DID, BOB_DID};

const NAMESPACE: &str = "integration-test";

#[derive(Clone)]
pub struct ProviderImpl {
    db: Surreal<Db>,
    blockstore: InMemoryBlockstore<64>,
    nats_client: async_nats::Client,
    pub keystore: KeystoreImpl,
}

impl Provider for ProviderImpl {}

impl ProviderImpl {
    pub async fn new() -> Result<Self> {
        let db = Surreal::new::<Mem>(()).await?;
        db.use_ns(NAMESPACE).use_db(ALICE_DID).await?;
        let blockstore = InMemoryBlockstore::<64>::new();
        let nats_client = async_nats::connect("demo.nats.io").await?;

        let provider = Self {
            db,
            blockstore,
            nats_client,
            keystore: KeystoreImpl::new(),
        };

        // load base protocol configuration for Alice and Bob
        let bytes = include_bytes!("./store/protocol.json");
        let config: Configure = serde_json::from_slice(bytes).expect("should deserialize");
        MessageStore::put(&provider, ALICE_DID, &config.clone().into()).await?;
        MessageStore::put(&provider, BOB_DID, &config.into()).await?;

        Ok(provider)
    }
}

#[derive(Debug, Deserialize)]
struct Entry {
    #[allow(dead_code)]
    id: RecordId,
}

impl DidResolver for ProviderImpl {
    async fn resolve(&self, url: &str) -> Result<Document> {
        serde_json::from_slice(include_bytes!("./store/did.json"))
            .map_err(|e| anyhow!(format!("issue deserializing document: {e}")))
    }
}
