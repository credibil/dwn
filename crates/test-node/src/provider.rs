#![allow(missing_docs)]
#![allow(unused_variables)]

//! # Provider
//!
//! Implementation of the `Provider` trait for testing and examples.

pub mod block_store;
mod event_stream;
pub mod key_store;

use anyhow::Result;
use blockstore::InMemoryBlockstore;
use dwn_node::provider::{
    DataStore, DidResolver, Document, EventLog, MessageStore, Provider, TaskStore,
};

#[derive(Clone)]
pub struct ProviderImpl {
    blockstore: InMemoryBlockstore<64>,
    pub nats_client: async_nats::Client,
}

impl ProviderImpl {
    pub async fn new() -> Result<Self> {
        Ok(Self {
            blockstore: InMemoryBlockstore::<64>::new(),
            nats_client: async_nats::connect("demo.nats.io").await?,
        })
    }
}

impl Provider for ProviderImpl {}
impl MessageStore for ProviderImpl {}
impl DataStore for ProviderImpl {}
impl EventLog for ProviderImpl {}
impl TaskStore for ProviderImpl {}

impl DidResolver for ProviderImpl {
    async fn resolve(&self, url: &str) -> Result<Document> {
        // let resolved =
        //     vercre_did::resolve(url, None, self.clone()).await.map_err(|e| anyhow!(e))?;
        // Ok(resolved.document.unwrap())

        // if url == &alice.did {
        //     serde_json::from_slice(include_bytes!("./provider/data/alice_did.json"))
        //         .map_err(|e| anyhow!(format!("issue deserializing document: {e}")))
        // } else {
        //     serde_json::from_slice(include_bytes!("./provider/data/bob_did.json"))
        //         .map_err(|e| anyhow!(format!("issue deserializing document: {e}")))
        // }

        println!("resolve: {}", url);

        Ok(Document::default())
    }
}
