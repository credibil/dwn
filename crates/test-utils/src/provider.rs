//! # Mock Provider

use std::sync::Arc;

use anyhow::Result;
use blockstore::{Blockstore as _, InMemoryBlockstore};
use cid::Cid;
use credibil_core::api::Client;
use credibil_dwn::event::{Event, Subscriber};
use credibil_dwn::provider::{BlockStore, EventStream, Resolver};
use credibil_proof::DocumentRequest;
use futures::stream::StreamExt;
use multihash_codetable::MultihashDigest;
use serde::{Deserialize, Serialize};

use crate::store::Store;

const RAW: u64 = 0x55;
const SUBJECT: &str = "events";
#[derive(Clone)]
pub struct Provider {
    blockstore: Arc<InMemoryBlockstore<64>>,
    pub nats_client: Arc<async_nats::Client>,
}

impl Provider {
    pub async fn new() -> Self {
        Self {
            blockstore: Arc::new(InMemoryBlockstore::<64>::new()),
            nats_client: Arc::new(
                async_nats::connect("demo.nats.io").await.expect("should connect"),
            ),
        }
    }
}

impl BlockStore for Provider {
    async fn put(&self, owner: &str, partition: &str, key: &str, block: &[u8]) -> Result<()> {
        let cid = Identitifier::new(owner, partition, key).to_cid()?;
        self.blockstore.put_keyed(&cid, block).await.map_err(Into::into)
    }

    async fn get(&self, owner: &str, partition: &str, key: &str) -> Result<Option<Vec<u8>>> {
        let cid = Identitifier::new(owner, partition, key).to_cid()?;
        let Some(bytes) = self.blockstore.get(&cid).await? else {
            return Ok(None);
        };
        Ok(Some(bytes))
    }

    async fn delete(&self, owner: &str, partition: &str, key: &str) -> Result<()> {
        let cid = Identitifier::new(owner, partition, key).to_cid()?;
        self.blockstore.remove(&cid).await?;
        Ok(())
    }

    async fn purge(&self, _owner: &str, _partition: &str) -> Result<()> {
        unimplemented!()
    }
}

#[derive(Serialize, Deserialize)]
struct Identitifier<'a> {
    owner: &'a str,
    partition: &'a str,
    key: &'a str,
}

impl<'a> Identitifier<'a> {
    fn new(owner: &'a str, partition: &'a str, key: &'a str) -> Self {
        Self {
            owner,
            partition,
            key,
        }
    }

    fn to_cid(&self) -> anyhow::Result<Cid> {
        let mut buf = Vec::new();
        ciborium::into_writer(self, &mut buf)?;
        let hash = multihash_codetable::Code::Sha2_256.digest(&buf);
        Ok(Cid::new_v1(RAW, hash))
    }
}

impl Resolver for Provider {
    async fn resolve(&self, url: &str) -> Result<Vec<u8>> {
        let request = DocumentRequest { url: url.to_string() };
        let document = Client::new(Store)
            .request(request)
            .owner("owner")
            .execute()
            .await
            .map(|r| r.0.clone())?;
        serde_json::to_vec(&document).map_err(|e| e.into())
    }
}

impl EventStream for Provider {
    /// Subscribe to a owner's event stream.
    async fn subscribe(&self, owner: &str) -> Result<Subscriber> {
        let subscriber = self.nats_client.subscribe(format!("{SUBJECT}.{owner}")).await?;
        let mapped = subscriber.map(|m| serde_json::from_slice::<Event>(&m.payload).unwrap());
        Ok(Subscriber::new(mapped))
    }

    /// Emits an event to a owner's event stream.
    async fn emit(&self, owner: &str, event: &Event) -> Result<()> {
        let bytes = serde_json::to_vec(event)?;
        self.nats_client.publish(format!("{SUBJECT}.{owner}"), bytes.into()).await?;
        Ok(())
    }
}
