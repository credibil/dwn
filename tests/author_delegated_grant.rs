//! Author Delegated Grant
//!
//! This test demonstrates how a web node owner can delegate permission to
//! another entity to perform an action on their behalf. In this case, Alice
//! grants Bob the ability to configure a protocol on her behalf.

use test_utils::store::ProviderImpl;
use vercre_dwn::permissions::GrantBuilder;
use vercre_dwn::protocols::{ConfigureBuilder, ProtocolDefinition, QueryBuilder};
use vercre_dwn::provider::KeyStore;
use vercre_dwn::service::{Message, Reply};
use vercre_dwn::{Interface, Method};

const ALICE_DID: &str = "did:key:z6Mkj8Jr1rg3YjVWWhg7ahEYJibqhjBgZt1pDCbT4Lv7D4HX";
const BOB_DID: &str = "did:key:z6Mkj8Jr1rg3YjVWWhg7ahEYJibqhjBgZt1pDCbT4Lv7D4HX";

#[tokio::test]
async fn configure() {
    let provider = ProviderImpl::new().await.expect("should create provider");
    let alice_keyring = provider.keyring(ALICE_DID).expect("should get Alice's keyring");
    let bob_keyring = provider.keyring(BOB_DID).expect("should get Alice's keyring");

    // ------------------------------
    // Alice grants Bob the ability to configure any protocol
    // ------------------------------
    let builder = GrantBuilder::new()
        .granted_to(BOB_DID)
        .request_id("grant_id_1")
        .description("Allow Bob to configure any protocol")
        .delegated(true)
        .scope(Interface::Protocols, Method::Configure, None);

    let grant_to_bob = builder.build(&alice_keyring).await.expect("should create grant");

    // ------------------------------
    // Bob configures the email protocol on Alice's behalf
    // ------------------------------
    let email_json = include_bytes!("protocols/email.json");
    let email_proto: ProtocolDefinition =
        serde_json::from_slice(email_json).expect("should deserialize");

    let configure = ConfigureBuilder::new()
        .definition(email_proto.clone())
        .delegated_grant(grant_to_bob)
        .build(&bob_keyring)
        .await
        .expect("should build");
    let message = Message::ProtocolsConfigure(configure);

    let reply = vercre_dwn::handle_message(ALICE_DID, message, provider.clone())
        .await
        .expect("should configure protocol");

    let Reply::ProtocolsConfigure(reply) = reply else {
        panic!("unexpected reply: {:?}", reply);
    };

    assert_eq!(reply.status.code, 202);

    // ------------------------------
    // Alice fetches the email protocol configured by Bob
    // ------------------------------
    let query = QueryBuilder::new()
        .filter(email_proto.protocol)
        .build(&alice_keyring)
        .await
        .expect("should build");
    let message = Message::ProtocolsQuery(query);

    let reply = vercre_dwn::handle_message(ALICE_DID, message, provider.clone())
        .await
        .expect("should find protocol");

    let Reply::ProtocolsQuery(reply) = reply else {
        panic!("unexpected reply: {:?}", reply);
    };

    assert_eq!(reply.status.code, 200);
}
