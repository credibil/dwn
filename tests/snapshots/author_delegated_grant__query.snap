---
source: tests/author_delegated_grant.rs
assertion_line: 84
expression: entry
snapshot_kind: text
---
status:
  code: 200
  detail: OK
entries:
  - descriptor:
      interface: Protocols
      method: Configure
      messageTimestamp: "[messageTimestamp]"
      definition:
        protocol: "http://email-protocol.xyz"
        published: true
        types:
          email:
            schema: "http://email"
            dataFormats:
              - text/plain
        structure:
          email:
            $actions:
              - who: anyone
                can:
                  - create
              - who: author
                of: email
                can:
                  - read
              - who: recipient
                of: email
                can:
                  - read
            email:
              $actions:
                - who: anyone
                  can:
                    - create
                - who: author
                  of: email/email
                  can:
                    - read
                - who: recipient
                  of: email/email
                  can:
                    - read
    authorization:
      signature:
        payload: "[payload]"
        signatures:
          - protected: eyJhbGciOiJFZERTQSIsInR5cCI6Imp3dCIsImtpZCI6ImRpZDprZXk6ejZNa2o4SnIxcmczWWpWV1doZzdhaEVZSmlicWhqQmdadDFwRENiVDRMdjdENEhYI3o2TWtqOEpyMXJnM1lqVldXaGc3YWhFWUppYnFoakJnWnQxcERDYlQ0THY3RDRIWCJ9
            signature: "[signature]"
